### DISCUSSION TITLE
blog/automatic-query-invalidation-after-mutations

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#105] [Announcements]
# blog/automatic-query-invalidation-after-mutations

Even though there is nothing built into React Query, it doesn't need a lot of code to implement automatic query invalidation in user-land thanks to the global cache callbacks.

https://tkdodo.eu/blog/automatic-query-invalidation-after-mutations#inject-comments

**Comment by kerematam** [comment]
Great post. Tbh I think Invalidation with mutationKey should come with default.

**Reply by TkDodo** [reply]
you can use `mutationKeys` to apply different mutation defaults for them. This hierarchy doesn't have to match what you do for queries. If we tie those things together, I'm sure someone will find a way to make those two interfere in an unexpected way :)

**Comment by bacarybruno** [comment]
This post comes at a very good moment! I was wondering if there was a way to do exactly this: invalidating queries without having to use the query client on mutations 👏

Is there a way to apply the same principle for optimistic updates? To avoid interacting with the cache on mutations.

**Reply by TkDodo** [reply]
I mean you probably can but it has the same problem for us to recommend anything: There are too many ways to do it, depending on your use-case. We can see that in the [swr api for optimistic updates](https://swr.vercel.app/docs/mutation.en-US#parameters). It takes `optimisticData`,  `revalidate`, `populateCache`, `rollbackOnError`, _and_ `throwOnError` as options to do optimistic updates, and I think it still doesn't cover all cases. That's a huge API surface ...

**Comment by lgenzelis** [comment]
The second approach, where you use `getQueryDefaults`, left me wondering if maybe I've been using tanstack query in a suboptimal way all this time. I have never set defaults for a query. Is that something you recommend? I usually just have a function that exposes the query to the whole app, and if I need to pass additional options to useQuery, I just put them as parameters in the function. E.g.:
```ts
export function useIssues(staleTime?: number) {
  return useQuery({
    queryFn: fetchIssues,
    queryKey: ['issues'],
    gcTime: 60_000, 
    ...(staleTime !== undefined && { staleTime }),
 });
}
```

**Reply by TkDodo** [reply]
I do it sometimes but it's mostly preference:

https://tkdodo.eu/blog/react-query-as-a-state-manager#bonus-using-setquerydefaults

**Reply by lgenzelis** [reply]
Got it, thank you! 🙌 I'd honestly recommend against that, because it'd mean a query could not behave as I expect it to just because it's matching a default set somewhere. I prefer having just two sources of truth: the `QueryClient` and the `useQuery` call itself. But ofc it's just my opinion :)

**Comment by vferak** [comment]
Thank you for this awesome blog! It's a great resource for learning Tanstack query. ❤️

Anyway, I just tried the invalidate everything approach on my project and i've found one issue with it: mutations that delete currently rendered entity.

In the onSuccess callback of the mutation function i try to navigate to different page ASAP, so that the current page does not rerender, but the query responsible for fetching the entity detail still goes through resulting in not found error before navigating away.

Is there an easy way how i can invalidate everything except some query keys? Or should i just stick with granular invalidations for delete mutations?

**Reply by TkDodo** [reply]
What I did is to invalidate everything, except if `invalidates` tags are passed. For some mutations, like `logout`, it just pass: `{ invalidates: [] }` to not invalidate anything.

**Reply by vferak** [reply]
Yes, i expected manual invalidation for these usecases would be the best. Thank you so much 👌

**Comment by fitimbytyqi** [comment]
Great and insightful post as always, one question I have is regarding trpc, I've noticed that trpc adds automatic query key's and a bit long ones.

Do we have to check whats the queryKey trpc has assigned from query devtools because we don't have control in queryKey manually?

**Reply by TkDodo** [reply]
there is a `getQueryKey` util from trpc: https://trpc.io/docs/client/react/getQueryKey

**Reply by TkDodo** [reply]
also, if you want to do invalidation, use the typed helpers: https://trpc.io/docs/client/react/useUtils

**Reply by fitimbytyqi** [reply]
That's great, thank you a lot!

**Comment by Sayemhaque** [comment]
what about that way?   
```queryClient.setQueryData(['course', id], (oldCourse: CourseViewModel) => {
        if (!oldCourse.curriculum) {
          return;
        }
        const CourseCurriculum = produce(oldCourse, course => {
          course.curriculum?.push({
            id: data.data.id,
            isCompleted: false,
            title: data.data.title,
            sequence: data.data.sequence,
            contents: [],
          });
        });

        return CourseCurriculum;
      });
```

**Reply by Sayemhaque** [reply]
```
queryClient.setQueryData(['course', id], (oldCourse: CourseViewModel) => {
        if (!oldCourse.curriculum) {
          return;
        }
        const CourseCurriculum = produce(oldCourse, course => {
          course.curriculum?.push({
            id: data.data.id,
            isCompleted: false,
            title: data.data.title,
            sequence: data.data.sequence,
            contents: [],
          });
        });

        return CourseCurriculum;
      });
```

**Comment by sregg** [comment]
Invalidating all queries can be more costly in React Native because previous screens stay mounted in the navigation stack and are still considered active for RQ.
This can lead to a lot more re-renders in the background and can lead to thread bottlenecks.

**Reply by sregg** [reply]
Nvm, there's already a fix for that in the docs
https://tanstack.com/query/latest/docs/framework/react/react-native#disable-re-renders-on-out-of-focus-screens

**Comment by AaronConlon** [comment]
Hi there,how to use `awaits` meta data to await some invalidations?

**Reply by TkDodo** [reply]
by calling `await queryClient.invalidateQueries` instead of just `queryClient.invalidateQueries`

**Comment by Lokendra-sinh** [comment]
Loved this post and got to know more about queries in general. My question is, is there a way we could invalidate query(s) via passing dynamic parameters so the query could use those to invalidate it???

**Reply by TkDodo** [reply]
you have access to the `mutationKey` and that can be dynamic, yes.

**Comment by wporoslo** [comment]
I have a table of emails sent and their statuses. After sending an email (mutation) I invalidate emails query and it returns data with a new entry of status `pending`. The email might take somewhere around ~15s to either error out or be delivered successfully and so I'd like to refetch the query twice after invalidation in 10s intervals to display accurate status fairly quickly and then return to the default behaviour. I've tried to using `useRef` inside custom `useQuery`, but to no avail; the code below refetches query immediately (I guess I may be just confused and the implementation is probably erroneous somehow):
```

export const useEmailsQuery = (data: { contactId: string }) => {
  const hasBeenInvalidated = useRef(false);
  const dataUpdateCountSinceLastInvalidation = useRef(0);

  return useQuery({
    queryKey: ['emails', data.contactId],
    queryFn: () => getEmails(data.contactId),
    placeholderData: keepPreviousData,
    // after succesful email sent we invalidate this query.
    // sending emails might take somewhere around ~15s and after this time
    // the status of the email changes, so we refetch 2 times after query invalidation
    // fairly quickly, then return to the default refetchInterval
    staleTime: 15000, // 15s
    refetchInterval: (query) => {
      if (query.state.isInvalidated) {
        hasBeenInvalidated.current = true;
        dataUpdateCountSinceLastInvalidation.current = 0;
      }

      if (
        hasBeenInvalidated.current &&
        dataUpdateCountSinceLastInvalidation.current < 2
      ) {
        dataUpdateCountSinceLastInvalidation.current++;
        return 1000 * 10; // 10s
      }

      // Reset after completing two 15-second intervals
      hasBeenInvalidated.current = false;
      dataUpdateCountSinceLastInvalidation.current = 0;

      return false
    },
  });
};

```
How would you go about achieving that refetching behaviour?

**Reply by TkDodo** [reply]
The idea is to just refetch until you get a status from the backend that isn't `pending` anymore ...

**Reply by wporoslo** [reply]
damn, so simple yet exactly what I needed, thx a lot!

**Comment by mihaiandrei97** [comment]
Awesome post. Thank you @TkDodo. A small question if you don't mind. With the first aproach:

```ts
const queryClient = new QueryClient({
  mutationCache: new MutationCache({
    onSuccess: (_data, _variables, _context, mutation) => {
      queryClient.invalidateQueries({
        queryKey: mutation.options.mutationKey,
      })
    },
  }),
})
```

How would you make a mutation that doesn't invalidate? Passing an empty array for `mutationKey` seems to still invalidate all.

```ts
export function useCreateTimerMutation() {
  return useMutation({
    mutationKey: ['none'],
    mutationFn: fn
  });
}
```

Thank you!

**Reply by mihaiandrei97** [reply]
So my solution was to just pass "none" as a string. What do you think?

**Reply by TkDodo** [reply]
set no `mutationKey` (it’s optional) and make a check in `onSuccess` before calling `invalidateQueries`

**Comment by Victor-Mestre-Sybo** [comment]
This is great! I have used this solution in my projects with TanStack Query. However, similar to the rest of the mutation callbacks, I'd adapt this so that `invalidates` also has access to the mutation variables, as they are sometimes needed to build the query keys that need to be invalidated.

**Comment by centraldogma99** [comment]
Hi @TkDodo, Thank you for the insightful article on automatic query invalidation. I have a question about the “To Await or not to Await” section.

In the example where you await specific query invalidation:

```javascript
useMutation({
  mutationFn: updateLabel,
  onSuccess: () => {
    return queryClient.invalidateQueries(
      { queryKey: ['labels'] },
      { cancelRefetch: false }
    )
  },
})
```

You mentioned earlier in the article that “Invalidation merely refetches all **active Queries** that it matches, and marks the rest as stale”.

This means that if the `['labels']` query is currently inactive (not mounted on screen), the `invalidateQueries` would complete immediately without actually triggering a refetch, right? So the mutation would also complete immediately, defeating the purpose of awaiting the refetch.

For example:

- User updates a label on a detail page
- The `['labels']` list query is inactive (user is not on the list page)
- `invalidateQueries` marks it as stale but doesn’t refetch
- The Promise resolves immediately without waiting for any actual refetch

One pattern I've been considering is separating the "invalidation" logic from the "awaiting" logic:
- Keep invalidation **without await** in mutation custom hooks
- Let components explicitly await(with something like `prefetchQuery`) what they need for their specific UI requirements

Or perhaps the example assumes that most of the queries that need to be invalidated are also active at the same time in most real-world scenarios? It would be helpful to clarify this assumption or mention the edge case for inactive queries.

Would love to hear your thoughts on how you handle this in practice!

**Reply by TkDodo** [reply]
> So the mutation would also complete immediately, defeating the purpose of awaiting the refetch.

yes, but you usually don’t control from within the mutation if `['labels']` matches any mounted queries or not. If it does, you await it. If not, await does nothing. That doesn’t defeat the purpose, that _is_ the purpose.

Your example shows why invalidation is good actually. If the user doesn’t navigate back to the list where labels are rendered, there’s no need to refetch them. They are not active now, so invalidation doesn’t do a refetch. If the user navigates back, they see stale content for a bit (stale-while-revalidate).

If you know that the user will be redirected back to a page where labels is used, e.g. because you do a programmatic redirect, I would maybe force a fetch here with `queryClient.refetchQueries` or by passing `refetchType: 'all'` to `invalidateQueries`.

But yes, if you have a route loader that would `await` the route requirements of “fresh data”, that’s also fine, and then you likely don’t ever need to await invalidations.

**Comment by oleksandrkovaliuk** [comment]
Hey, thanks for the awesome article!

When I was trying to implement global invalidation of all active queries in the mutation cache callback, I faced a weird but possibly expected behaviour that i would love to discuss with you!!

### Use case

I implemented a **global invalidation mechanism** as described, by calling:

```ts
queryClient.invalidateQueries()
``` 

inside a mutation cache callback.

I also wanted to explicitly mark some queries as “untouched” during global invalidation.
As suggested, i set their ```staleTime``` to ```Infinity``` and then use the ```predicate``` option in ```invalidateQueries``` to filter out queries whose ```staleTime``` is not ```Infinity```.

Here’s what I’ve end up with:
```ts
function nonStaticQueries(query: Query) {
  const defaultStaleTime =
    queryClient.getQueryDefaults(query.queryKey).staleTime ?? 0;

  const staleTimes = query.observers
    .map((observer) => observer.options.staleTime)
    .filter((staleTime) => staleTime !== undefined);

  const staleTime =
    query.getObserversCount() > 0
      ? Math.min(...(staleTimes as unknown as Array<number>))
      : defaultStaleTime;

  return staleTime !== Number.POSITIVE_INFINITY;
}

const mutationCache = new MutationCache({
  onSuccess: async (_data, _variables, _context, mutation) => {
    const successMessage = mutation.meta?.successMessage;

    if (successMessage) {
      toast.success(successMessage);
    }

    // @reference https://tkdodo.eu/blog/automatic-query-invalidation-after-mutations
    void queryClient.invalidateQueries({
      predicate: nonStaticQueries,
    });
  },
});
```

### Issue

Keeping it as is will never return ```true``` from the ```predicate```.
I think it’s because we’ve passed an object of options to invalidateQueries, and now it expects ```queryKey``` to be in it.
And it defaults ```queryKey``` to be ```undefined``` i assume. 

Which makes ```query.observers``` inside of ```nonStaticQueries``` to always return  [undefined] which ends up filtered to [] and then with [] as ```staleTimes``` in our expression here

```ts
const staleTime =
    query.getObserversCount() > 0
      ? Math.min(...(staleTimes as unknown as Array<number>))
      : defaultStaleTime;

  return staleTime !== Number.POSITIVE_INFINITY;
```

we will have ```staleTime``` being equal to ```Math.min(...[]) - Infinity``` which is equal to  ```Number.POSITIVE_INFINITY``` as result we are getting ```false``` for all active queries.

The solution i found to this issue is to set ```{exact:false}``` as options in invalidateQueries which will skip keys default matching algorithm and revalidate all active queries 

### Observation
On the other hand, ```queryClient.invalidateQueries()``` with no options provided correctly matches and invalidates all active queries in the cache.

### Questions 
1.Why does ```queryClient.invalidateQueries()``` behave differently than ```queryClient.invalidateQueries({ predicate: nonStaticQueries })``` even though we didn’t specify any value for ```exact``` how it should match keys?
2.What are your general thoughts on using a predicate with exact: false
without defining an explicit ```queryKey```, as you did with ```invalidates``` meta?
3.Wouldn’t it make sense to note this somewhere in the article?
(Maybe I’m the only one who faced this, but just curious and suggesting!)

Huge thanks for all the work you’re doing for the community !! Have a great day !

**Reply by oleksandrkovaliuk** [reply]
### Additional investigation

Issue was not fixed by setting `exact:false` as mentioned above.

The reason for nonStaticQuery to throw away queries with no `staleTime` set was condition for `staleTime`

See example and read comment below

```ts
function nonStaticQueries(query: Query) {
  const defaultStaleTime =
    queryClient.getQueryDefaults(query.queryKey).staleTime ?? 0;

  const staleTimes = query.observers
    .map((observer) => observer.options.staleTime)
    .filter((staleTime) => staleTime !== undefined);

  // change query.getObserversCount() to staleTimes?.length
  // since we can have observers but no staleTimes set in the query itself
  // which will result in staleTime being Infinity
  // and then nonStaticQueries will always return false for queries with no staleTimes set
  const staleTime = staleTimes?.length
    ? Math.min(...(staleTimes as unknown as Array<number>))
    : defaultStaleTime;

  console.group(query.queryKey);
  console.log("Infinity check-", staleTime !== Number.POSITIVE_INFINITY);
  console.log("staleTimes-", staleTimes);
  console.log("query.queryKey-", query.queryKey);
  console.groupEnd();

  return staleTime !== Number.POSITIVE_INFINITY;
}
```

With implementation above it works!

But would love to hear your thoughts @TkDodo
```ts
void queryClient.invalidateQueries({
        predicate: nonStaticQueries,
      });
```

Thanks!

**Reply by TkDodo** [reply]
we’ve since released `staleTime: 'static'`, which will _never_ be refetched and not affected by `queryClient.invalidateQueries()`. This sounds like what you want here

**Reply by TkDodo** [reply]
I’ve updated the blogpost 👍

**Reply by oleksandrkovaliuk** [reply]
Lovely , sorry i missed this option , thanks !
